lib/customisation.nix

```
@@ -661,4 +662,186 @@ rec {
+
+    # Inputs
+
+    `extendMkDerivation`-specific configurations
+    : `inheritFunctionArgs`: Whether to inherit `__functionArgs` from the base build helper (default to `true`)
+    : `modify`: Function to modify the result derivation (default to `lib.id`)
+
+    `mkDerivationBase`
+    : Base build helper, the `mkDerivation`-like build helper to extend
+
+    `attrsOverlay`
+    : An overlay of attribute set, like the one taken by [overrideAttrs](#sec-pkg-overrideAttrs).
+    : It is the implementation detail of the result build helper.
+
+    # Type
+
+    ```
+    extendMkDerivation ::
+      {
+        inheritFunctionArgs :: Bool,
+        modify :: a -> a,
+      }
+      -> ((FixedPointArgs | AttrSet) -> a)
+      -> (AttrSet -> AttrSet -> AttrSet)
+      -> (FixedPointArgs | AttrSet) -> a
+
+    FixedPointArgs = AttrSet -> AttrSet
+    a = Derivation when defining a build helper
+    ```
+
+    # Examples
+
+    :::{.example}
+    ## `lib.customisation.extendMkDerivation` usage example
+    ```nix-shell
+    mkLocalDerivation = lib.extendMkDerivation { } pkgs.stdenv.mkDerivation (finalAttrs:
+      args@{ preferLocalBuild ? true, allowSubstitute ? false, ... }:
+      { inherit preferLocalBuild allowSubstitute; })
+
+    mkLocalDerivation.__functionArgs
+    => { allowSubstitute = true; preferLocalBuild = true; }
+
+    mkLocalDerivation { inherit (pkgs.hello) pname version src; }
+    => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+    mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; })
+    => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+    (mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; passthru = { foo = "a"; bar = "${finalAttrs.passthru.foo}b"; } })).bar
+    => "ab"
+    ```
+    :::
+
+    :::{.note}
+    If *modify* is specified,
+    it should take care of existing attributes that perform overriding
+    (e.g., [`overrideAttrs`](#sec-pkg-overrideAttrs))
+    to ensure that the overriding functionality of the result derivation
+    work as expected.
+    Modifications that breaks the overriding include
+    direct [attribute set update](https://nixos.org/manual/nix/stable/language/operators#update)
+    and [`lib.extendDerivation`](#function-library-lib.customisation.extendDerivation).
+    :::
+  */
+  extendMkDerivation =
+    {
+      modify ? id,
+      inheritFunctionArgs ? true,
+    }:
+    mkDerivationBase: attrsOverlay:
+    setFunctionArgs
+      # Adds the fixed-point style support.
+      (fpargs: modify ((mkDerivationBase fpargs).overrideAttrs attrsOverlay))
+      # Add __functionArgs
+      (
+        # Inherit the __functionArgs from the base build helper
+        optionalAttrs inheritFunctionArgs (functionArgs mkDerivationBase)
+        # Recover the __functionArgs from the derived build helper
+        // functionArgs (attrsOverlay { })
+      )
```

```suggestion
      (
        # Inherit the __functionArgs from the base build helper
        functionArgs (attrsOverlay { })
        # Recover the __functionArgs from the derived build helper
        // optionalAttrs inheritFunctionArgs (functionArgs mkDerivationBase)
      )
```

Note that `overlay-a` is before `overlay-b` (the reverse order of `adaptMkDerivation`).
```nix
extendMkDerivation { } (extendMkDerivation { } stdenv.mkDerivation overlay-a) overlay-b
```

---

> should that argument foo be removed from the result __functionArgs?

Ideally, yes. However, there's no programmatic way to achieve such removal.

---

> > should that argument foo be removed from the result __functionArgs?
> 
> Ideally, yes. However, there's no programmatic way to achieve such removal.

Update: it would be possible if we make `adaptMkDerivaiton` take an argument `removedAttrs` and then takes a function that returns the attributes to update. The usage of such interface would be

```nix
{
  mkLocalDerivation = adaptMkDerivation
    {
      removedAttrs = [ "specialArgs" ];
    }
    stdenv.mkDerivation
    (
      finalAttrs:
      {
        preferLocalBuild ? true,
        allowSubstitutes ? false,
        specialArg ? (_: false),
        ...
      }@args:
      {
        inherit preferLocalBuild allowSubstitutes;
        passthru = { inherit specialArgs; } // args.passthru or { };
      }
    );
}
```

Such interface is similar to `extendMkDerivation`:


```nix
{
  mkLocalDerivation = extendMkDerivation
    { }
    stdenv.mkDerivation
    (
      finalAttrs:
      {
        preferLocalBuild ? true,
        allowSubstitutes ? false,
        ...
      }@args:
      {
        inherit preferLocalBuild allowSubstitutes;
      }
    );
}
```